Controller methods that take one or more IDs should take a single list arg instead of variable args

This will make usage more convenient for dronefly's wrappers.

Example traceback from Ben:
```python
[2023-03-03 09:33:03] [ERROR] red: Exception in command 'event leave'
Traceback (most recent call last):
  File "/home/synrg/.cache/pypoetry/virtualenvs/dronefly-xT1cyIDA-py3.9/lib/python3.9/site-packages/discord/ext/commands/core.py", line 85, in wrapped
    ret = await coro(*args, **kwargs)
  File "/home/synrg/work/dronefly/inatcog/commands/event.py", line 122, in event_leave
    await self._event_action(ctx, "leave", abbrev, user)
  File "/home/synrg/work/dronefly/inatcog/commands/event.py", line 74, in _event_action
    response = await self.api.update_project_users(
  File "/home/synrg/work/dronefly/inatcog/api.py", line 397, in update_project_users
    response = await inat.projects.delete_users(project_id, user_ids)
  File "/home/synrg/.cache/pypoetry/virtualenvs/dronefly-xT1cyIDA-py3.9/lib/python3.9/site-packages/forge/_revision.py", line 328, in inner
    return callable(*mapped.args, **mapped.kwargs)
  File "/home/synrg/.cache/pypoetry/virtualenvs/dronefly-xT1cyIDA-py3.9/lib/python3.9/site-packages/pyinaturalist/controllers/project_controller.py", line 68, in delete_users
    response = self.client.request(delete_project_users, project_id, user_ids, **params)
  File "/home/synrg/.cache/pypoetry/virtualenvs/dronefly-xT1cyIDA-py3.9/lib/python3.9/site-packages/pyinaturalist/client.py", line 183, in request
    return request_function(*args, **kwargs)
TypeError: delete_project_users() got multiple values for argument 'user_ids'
```

--- a/HISTORY.md
+++ b/HISTORY.md
@@ -1,5 +1,7 @@
 # History
 
+## 0.19.0 (Unreleased)
+
 ## 0.18.0 (2023-02-27)
 
 ### New Endpoints

--- a/pyinaturalist/__init__.py
+++ b/pyinaturalist/__init__.py
@@ -1,6 +1,6 @@
 # flake8: noqa: F401, F403
 # isort: skip_file
-__version__ = '0.18.1'
+__version__ = '0.19.0'
 
 from pyinaturalist.auth import get_access_token
 from pyinaturalist.client import iNatClient

--- a/pyinaturalist/controllers/observation_controller.py
+++ b/pyinaturalist/controllers/observation_controller.py
@@ -6,8 +6,10 @@ from pyinaturalist.constants import (
     HistogramResponse,
     IntOrStr,
     ListResponse,
+    MultiInt,
 )
 from pyinaturalist.controllers import BaseController
+from pyinaturalist.converters import ensure_list
 from pyinaturalist.docs import document_controller_description, document_controller_params
 from pyinaturalist.models import (
     ControlledTermCounts,
@@ -48,7 +50,7 @@ class ObservationController(BaseController):
         return self.from_ids(observation_id, **kwargs).one()
 
     @document_controller_description(get_observations_by_id)
-    def from_ids(self, *observation_ids: int, **params) -> Paginator[Observation]:
+    def from_ids(self, observation_ids: MultiInt, **params) -> Paginator[Observation]:
         """
         Args:
             observation_ids: One or more observation IDs
@@ -57,7 +59,7 @@ class ObservationController(BaseController):
             get_observations_by_id,
             Observation,
             cls=IDPaginator,
-            ids=observation_ids,
+            ids=ensure_list(observation_ids),
             ids_per_request=IDS_PER_REQUEST,
             **params,
         )
@@ -120,13 +122,13 @@ class ObservationController(BaseController):
         response = self.client.request(create_observation, auth=True, **params)
         return Observation.from_json(response)
 
-    def delete(self, *observation_ids: int, **params):
+    def delete(self, observation_ids: MultiInt, **params):
         """Delete observations
 
         Args:
             observation_ids: One or more observation IDs
         """
-        for observation_id in observation_ids:
+        for observation_id in ensure_list(observation_ids):
             self.client.request(
                 delete_observation,
                 auth=True,

--- a/pyinaturalist/controllers/place_controller.py
+++ b/pyinaturalist/controllers/place_controller.py
@@ -1,7 +1,8 @@
 from typing import Optional
 
-from pyinaturalist.constants import IntOrStr
+from pyinaturalist.constants import MultiIntOrStr
 from pyinaturalist.controllers import BaseController
+from pyinaturalist.converters import ensure_list
 from pyinaturalist.docs import document_controller_params
 from pyinaturalist.models import Place
 from pyinaturalist.paginator import AutocompletePaginator, Paginator
@@ -15,13 +16,15 @@ class PlaceController(BaseController):
         """Get a single place by ID"""
         return self.from_ids(place_id, **kwargs).one()
 
-    def from_ids(self, *place_ids: IntOrStr, **params) -> Paginator[Place]:
+    def from_ids(self, place_ids: MultiIntOrStr, **params) -> Paginator[Place]:
         """Get places by ID
 
         Args:
             place_ids: One or more place IDs
         """
-        return self.client.paginate(get_places_by_id, Place, place_id=place_ids, **params)
+        return self.client.paginate(
+            get_places_by_id, Place, place_id=ensure_list(place_ids), **params
+        )
 
     @document_controller_params(get_places_autocomplete)
     def autocomplete(self, q: Optional[str] = None, **params) -> Paginator[Place]:

--- a/pyinaturalist/controllers/project_controller.py
+++ b/pyinaturalist/controllers/project_controller.py
@@ -1,7 +1,8 @@
 from typing import Optional
 
-from pyinaturalist.constants import IntOrStr, ListResponse
+from pyinaturalist.constants import IntOrStr, ListResponse, MultiInt, MultiIntOrStr
 from pyinaturalist.controllers import BaseController
+from pyinaturalist.converters import ensure_list
 from pyinaturalist.docs import document_controller_params
 from pyinaturalist.models import Project
 from pyinaturalist.paginator import Paginator
@@ -22,7 +23,7 @@ class ProjectController(BaseController):
         """Get a single project by ID"""
         return self.from_ids(project_id, **kwargs).one()
 
-    def from_ids(self, *project_ids: IntOrStr, **params) -> Paginator[Project]:
+    def from_ids(self, project_ids: MultiIntOrStr, **params) -> Paginator[Project]:
         """Get projects by ID
 
         Args:
@@ -34,7 +35,9 @@ class ProjectController(BaseController):
     def search(self, **params) -> Paginator[Project]:
         return self.client.paginate(get_projects, Project, **params)
 
-    def add_observations(self, project_id: int, *observation_ids: int, **params) -> ListResponse:
+    def add_observations(
+        self, project_id: int, observation_ids: MultiInt, **params
+    ) -> ListResponse:
         """Add an observation to a project
 
         Args:
@@ -42,7 +45,7 @@ class ProjectController(BaseController):
             observation_ids: One or more observation IDs to add
         """
         responses = []
-        for observation_id in observation_ids:
+        for observation_id in ensure_list(observation_ids):
             response = self.client.request(
                 add_project_observation,
                 project_id=project_id,
@@ -55,15 +58,17 @@ class ProjectController(BaseController):
 
     @document_controller_params(update_project)
     def update(self, project_id: IntOrStr, **params) -> Project:
-        response = self.client.request(update_project, project_id, **params)
+        response = self.client.request(update_project, project_id, auth=True, **params)
         return Project.from_json(response)
 
     @document_controller_params(add_project_users)
-    def add_users(self, project_id: IntOrStr, *user_ids: int, **params) -> Project:
-        response = self.client.request(add_project_users, project_id, user_ids, **params)
+    def add_users(self, project_id: IntOrStr, user_ids: MultiInt, **params) -> Project:
+        response = self.client.request(add_project_users, project_id, user_ids, auth=True, **params)
         return Project.from_json(response)
 
     @document_controller_params(delete_project_users)
-    def delete_users(self, project_id: IntOrStr, *user_ids: int, **params) -> Project:
-        response = self.client.request(delete_project_users, project_id, user_ids, **params)
+    def delete_users(self, project_id: IntOrStr, user_ids: MultiInt, **params) -> Project:
+        response = self.client.request(
+            delete_project_users, project_id, user_ids, auth=True, **params
+        )
         return Project.from_json(response)

--- a/pyinaturalist/controllers/taxon_controller.py
+++ b/pyinaturalist/controllers/taxon_controller.py
@@ -2,7 +2,9 @@ from typing import Optional
 
 from attr import fields_dict
 
+from pyinaturalist.constants import MultiInt
 from pyinaturalist.controllers import BaseController
+from pyinaturalist.converters import ensure_list
 from pyinaturalist.docs import document_controller_params
 from pyinaturalist.models import Taxon
 from pyinaturalist.paginator import IDPaginator, Paginator
@@ -18,7 +20,7 @@ class TaxonController(BaseController):
         """Get a single taxon by ID"""
         return self.from_ids(taxon_id, **kwargs).one()
 
-    def from_ids(self, *taxon_ids: int, **params) -> Paginator[Taxon]:
+    def from_ids(self, taxon_ids: MultiInt, **params) -> Paginator[Taxon]:
         """Get taxa by ID
 
         Args:
@@ -28,7 +30,7 @@ class TaxonController(BaseController):
             get_taxa_by_id,
             Taxon,
             cls=IDPaginator,
-            ids=taxon_ids,
+            ids=ensure_list(taxon_ids),
             ids_per_request=IDS_PER_REQUEST,
             **params
         )

--- a/pyinaturalist/controllers/user_controller.py
+++ b/pyinaturalist/controllers/user_controller.py
@@ -1,7 +1,8 @@
 from typing import List, Optional
 
-from pyinaturalist.constants import IntOrStr
+from pyinaturalist.constants import MultiIntOrStr
 from pyinaturalist.controllers import BaseController
+from pyinaturalist.converters import ensure_list
 from pyinaturalist.docs import document_controller_params
 from pyinaturalist.models import User
 from pyinaturalist.paginator import IDPaginator, Paginator
@@ -15,7 +16,7 @@ class UserController(BaseController):
         """Get a single user by ID"""
         return self.from_ids(user_id, **kwargs).one()
 
-    def from_ids(self, *user_ids: IntOrStr, **params) -> Paginator[User]:
+    def from_ids(self, user_ids: MultiIntOrStr, **params) -> Paginator[User]:
         """Get users by ID
 
         Example:
@@ -30,7 +31,9 @@ class UserController(BaseController):
         Args:
             user_ids: One or more project IDs
         """
-        return self.client.paginate(get_user_by_id, User, cls=IDPaginator, ids=user_ids, **params)
+        return self.client.paginate(
+            get_user_by_id, User, cls=IDPaginator, ids=ensure_list(user_ids), **params
+        )
 
     # TODO: Wrap in paginator?
     @document_controller_params(get_users_autocomplete)

--- a/pyinaturalist/session.py
+++ b/pyinaturalist/session.py
@@ -413,7 +413,7 @@ def get_refresh_params(endpoint) -> Dict:
             REFRESH_LIMITER.try_acquire(bucket)
             break
         except BucketFullException as e:
-            seconds = int(e.meta_info["remaining_time"])
+            seconds = int(e.meta_info['remaining_time'])
             logger.debug(f'{bucket} cannot be refreshed again for {seconds} seconds')
             v += 1
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "pyinaturalist"
-version = "0.18.1"
+version = "0.19.0"
 description = "iNaturalist API client for python"
 authors = ["Nicolas Noé <nicolas@niconoe.eu>", "Jordan Cook <Jordan.Cook@pioneer.com>"]
 license = "MIT"

--- a/test/controllers/test_place_controller.py
+++ b/test/controllers/test_place_controller.py
@@ -21,7 +21,7 @@ def test_from_ids(requests_mock):
         json={'results': [j_place_1, j_place_2], 'total_results': 2},
         status_code=200,
     )
-    results = iNatClient().places.from_ids(89191, 67591).all()
+    results = iNatClient().places.from_ids([89191, 67591]).all()
     place = results[0]
 
     assert len(results) == 2 and isinstance(place, Place)

--- a/test/controllers/test_project_controller.py
+++ b/test/controllers/test_project_controller.py
@@ -96,7 +96,7 @@ def test_add_observation(get_access_token, requests_mock):
         status_code=200,
     )
     observation_ids = 5678, 9012
-    iNatClient().projects.add_observations(1234, *observation_ids)
+    iNatClient().projects.add_observations(1234, observation_ids)
     get_access_token.assert_called()
 
 

--- a/test/controllers/test_user_controller.py
+++ b/test/controllers/test_user_controller.py
@@ -28,7 +28,7 @@ def test_from_ids(requests_mock):
         status_code=200,
     )
 
-    results = iNatClient().users.from_ids(1, 2).all()
+    results = iNatClient().users.from_ids([1, 2]).all()
     assert len(results) == 2 and isinstance(results[0], User)
     assert results[0].id == 1
 

